File: cmd/shen-derive-ts/core/match.ts
// Pattern matcher for shen-derive. Ported from shen-derive/core/match.go.
//
// Matches a pattern (represented as a `Sexpr`) against a runtime `Value`
// produced by the evaluator. The `Value` union and its constructors live
// in `./eval.ts`; this module re-exports the handful of constructors
// pattern-matching consumers typically need so existing test imports
// keep working.

import type { Sexpr } from "./sexpr.ts";
import {
  type Value,
  type ListVal,
  valEqual,
} from "./eval.ts";

export {
  type Value,
  type IntVal,
  type FloatVal,
  type BoolVal,
  type StringVal,
  type ListVal,
  intVal,
  floatVal,
  boolVal,
  stringVal,
  listVal,
  valEqual,
} from "./eval.ts";

// --- Match result ---

export type MatchResult =
  | { kind: "matched"; bindings: Map<string, Value> }
  | { kind: "miss" }
  | { kind: "error"; msg: string };

// --- Match entry point ---
//
// Supported patterns:
//   - symbol "_"                — wildcard, binds nothing
//   - symbol "Uppercase..."     — variable binding
//   - symbol "nil"              — matches an empty list
//   - int / float / bool / str  — literal; matches by valEqual
//   - (cons head tail)          — cons pattern; matches a non-empty list
//
// Fixed-length list patterns like `[A B]` desugar to
// `(cons A (cons B nil))`, so the cons case is the only list form needed.
export function match(pat: Sexpr, v: Value): MatchResult {
  const bindings = new Map<string, Value>();
  const err = matchInto(pat, v, bindings);
  if (err === null) return { kind: "matched", bindings };
  if (err === NO_MATCH) return { kind: "miss" };
  return { kind: "error", msg: err };
}

// Internal sentinel distinct from any real error string.
const NO_MATCH = "\x00__no_match__\x00";

function matchInto(pat: Sexpr, v: Value, out: Map<string, Value>): string | null {
  if (pat.kind === "atom") return matchAtom(pat, v, out);
  return matchList(pat, v, out);
}

function isUpper(ch: string): boolean {
  return ch.length > 0 && ch >= "A" && ch <= "Z";
}

function matchAtom(
  p: Extract<Sexpr, { kind: "atom" }>,
  v: Value,
  out: Map<string, Value>,
): string | null {
  switch (p.atomKind) {
    case "symbol": {
      if (p.val === "_") return null;
      if (p.val === "nil") {
        if (v.kind === "list" && v.elems.length === 0) return null;
        return NO_MATCH;
      }
      if (p.val.length > 0 && isUpper(p.val[0]!)) {
        // Variable binding. Last-one-wins for repeated names (matches Go).
        out.set(p.val, v);
        return null;
      }
      return `unsupported symbol pattern "${p.val}" (only _, nil, and uppercase vars are allowed)`;
    }
    case "int": {
      const n = Number.parseInt(p.val, 10);
      if (Number.isNaN(n)) {
        return `bad int literal in pattern: ${p.val}`;
      }
      const lit: Value = { kind: "int", val: n };
      return valEqual(lit, v) ? null : NO_MATCH;
    }
    case "float": {
      const f = Number.parseFloat(p.val);
      if (Number.isNaN(f)) {
        return `bad float literal in pattern: ${p.val}`;
      }
      const lit: Value = { kind: "float", val: f };
      return valEqual(lit, v) ? null : NO_MATCH;
    }
    case "bool": {
      const want = p.val === "true";
      if (v.kind !== "bool" || v.val !== want) return NO_MATCH;
      return null;
    }
    case "string": {
      if (v.kind !== "string" || v.val !== p.val) return NO_MATCH;
      return null;
    }
  }
  return `unsupported atom kind in pattern: ${(p as { atomKind: string }).atomKind}`;
}

function matchList(
  p: Extract<Sexpr, { kind: "list" }>,
  v: Value,
  out: Map<string, Value>,
): string | null {
  const head = p.elems[0];
  const isCons =
    p.elems.length === 3 &&
    head !== undefined &&
    head.kind === "atom" &&
    head.atomKind === "symbol" &&
    head.val === "cons";
  if (!isCons) {
    return `unsupported list pattern (only cons patterns are supported)`;
  }
  if (v.kind !== "list" || v.elems.length === 0) return NO_MATCH;
  // head pattern -> first element
  const headErr = matchInto(p.elems[1]!, v.elems[0]!, out);
  if (headErr !== null) return headErr;
  // tail pattern -> rest of list as a ListVal
  const tail: ListVal = { kind: "list", elems: v.elems.slice(1) };
  return matchInto(p.elems[2]!, tail, out);
}

File: shen-derive/core/match.go
package core

import (
	"fmt"
	"strconv"
)

// Match tries to match a Shen-style pattern (an Sexpr) against a runtime
// Value. On success it returns the variable bindings produced by the match.
// On structural mismatch it returns (nil, false, nil). On an unrecognized
// pattern shape it returns a non-nil error.
//
// Supported patterns:
//   - Atom{Sym "_"}                     — wildcard, binds nothing
//   - Atom{Sym "Uppercase"}             — variable binding
//   - Atom{Sym "nil"}                   — matches an empty list
//   - Atom{Int/Float/Bool/String}       — literal; matches by valEqual
//   - List{cons, head, tail}            — cons pattern; matches a non-empty
//                                         list, binds head and tail recursively
//
// Fixed-length list patterns like `[A B]` desugar to
// `(cons A (cons B nil))`, so the cons case is the only list form needed.
func Match(pat Sexpr, v Value) (map[string]Value, bool, error) {
	env, ok, err := MatchEnv(pat, v, nil)
	if !ok {
		return nil, false, err
	}
	bindings := map[string]Value{}
	for e := env; e != nil; e = e.parent {
		if _, seen := bindings[e.name]; !seen {
			bindings[e.name] = e.val
		}
	}
	return bindings, true, nil
}

// MatchEnv is Match that binds the pattern's variables by extending env
// instead of building a map, which is what an evaluator needs and
// allocates far less. On a structural mismatch it returns (nil, false,
// nil).
func MatchEnv(pat Sexpr, v Value, env *Env) (*Env, bool, error) {
	env, err := matchInto(pat, v, env)
	if err != nil {
		if err == errNoMatch {
			return nil, false, nil
		}
		return nil, false, err
	}
	return env, true, nil
}

// sentinel error used to signal a structural non-match (as distinct from a
// malformed pattern, which returns a real error).
var errNoMatch = fmt.Errorf("no match")

func matchInto(pat Sexpr, v Value, env *Env) (*Env, error) {
	switch p := pat.(type) {
	case *Atom:
		return matchAtom(p, v, env)
	case *List:
		return matchList(p, v, env)
	}
	return nil, fmt.Errorf("unsupported pattern node: %T", pat)
}

func matchAtom(p *Atom, v Value, env *Env) (*Env, error) {
	switch p.Kind {
	case AtomSymbol:
		switch {
		case p.Val == "_":
			return env, nil
		case p.Val == "nil":
			if lv, ok := v.(ListVal); ok && len(lv) == 0 {
				return env, nil
			}
			return nil, errNoMatch
		case isUpper(p.Val[0]):
			// Variable binding. Repeated uppercase names in the same clause
			// would require consistency; for now, last-one-wins. The callers
			// of Match don't use repeated names, so we don't enforce it.
			return env.Extend(p.Val, v), nil
		default:
			return nil, fmt.Errorf("unsupported symbol pattern %q "+
				"(only _, nil, and uppercase vars are allowed)", p.Val)
		}
	case AtomInt, AtomFloat:
		// Compare numerically without boxing the literal, as valEqual
		// would: IntVal(1) matches the pattern 1.0 and vice versa.
		want, err := strconv.ParseFloat(p.Val, 64)
		if err != nil {
			return nil, fmt.Errorf("bad numeric literal in pattern: %s", p.Val)
		}
		if got, ok := asNum(v); ok && got == want {
			return env, nil
		}
		return nil, errNoMatch
	case AtomBool:
		want := p.Val == "true"
		bv, ok := v.(BoolVal)
		if !ok || bool(bv) != want {
			return nil, errNoMatch
		}
		return env, nil
	case AtomString:
		sv, ok := v.(StringVal)
		if !ok || string(sv) != p.Val {
			return nil, errNoMatch
		}
		return env, nil
	}
	return nil, fmt.Errorf("unsupported atom kind in pattern: %v", p.Kind)
}

func matchList(p *List, v Value, env *Env) (*Env, error) {
	if HeadSym(p) != "cons" || len(p.Elems) != 3 {
		return nil, fmt.Errorf("unsupported list pattern %q "+
			"(only cons patterns are supported)", p.String())
	}
	lv, ok := v.(ListVal)
	if !ok || len(lv) == 0 {
		return nil, errNoMatch
	}
	// head pattern → first element
	env, err := matchInto(p.Elems[1], lv[0], env)
	if err != nil {
		return nil, err
	}
	// tail pattern → rest of list as a ListVal
	return matchInto(p.Elems[2], lv[1:], env)
}

func isUpper(b byte) bool { return b >= 'A' && b <= 'Z' }
